### PR TITLE
Refactors DUBBD-AWS IaC and introduces testing

### PR DESCRIPTION
**Pull request by UncleGedd**
Addresses: https://github.com/defenseunicorns/uds-package-dubbd/issues/277

- refactors the Loki IaC into it's own `zarf.yaml` and imports its components into the DUBBD-AWS `zarf.yaml` so we can now test the Loki IaC in isolation
- adds Terratest to `aws/loki`
- adds workflow for Terratest and only runs when files in the `aw/loki` directory are modified
- makes EKS e2e tests run nightly or on modifications to certain files

### CODE REVIEW

File: aws/loki/main.tf [hunk 1]
```diff
@@ -26,20 +26,21 @@ locals {
   oidc_arn                  = "arn:${data.aws_partition.current.partition}:iam::${data.aws_caller_identity.current.account_id}:oidc-provider/${local.oidc_url_without_protocol}"
   generate_kms_key          = var.create_kms_key == true ? 1 : 0
   kms_key_arn               = var.kms_key_arn == null ? module.generate_kms[0].kms_key_arn : var.kms_key_arn
+  create_irsa               = var.create_irsa == true ? true : false
```
Comment (justin-o12) [inline aws/loki/main.tf:29]: Is this used?  It looks like you reference the `var` below in the S3 module.

File: aws/loki/variables.tf [hunk 2]
```diff
@@ -13,6 +13,11 @@ variable "key_alias" {
   default     = "bigbang-loki"
 }
 
+variable "create_irsa" {
+  description = "Create IRSA roles and plicy"
```
Comment (justin-o12) [inline aws/loki/variables.tf:17]: nit: typo

**Review by bdfinst — DISMISSED** [review]
LGTM

File: .github/workflows/test-loki-bucket.yaml [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+
+      - name: Upload test log
+        uses: actions/upload-artifact@v2
+        if: always()
+        with:
+          name: test-log
+          path: /tmp/gotest.log
+          if-no-files-found: error
+
+  test-loki-bucket-zarf:
```
Comment (lucasrod16) [inline .github/workflows/test-loki-bucket.yaml:48]: We may want to add a dependency on passing tests for the zarf package/deployment steps to run:

```yaml
    needs: test-loki-bucket-tf
```

File: .github/workflows/test-loki-bucket.yaml [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+        uses: defenseunicorns/setup-zarf@main
+        with:
+          # renovate: datasource=github-tags depName=defenseunicorns/zarf versioning=semver
+          version: v0.27.1
+          download-init-package: ${{ inputs.download-init-package }}
+
+      - name: Build Loki Zarf Package
+        working-directory: aws/loki
+        run: |
+          zarf package create
```
Comment (lucasrod16) [inline .github/workflows/test-loki-bucket.yaml:67]: We will need to add the `--confirm` flag here

File: .github/workflows/test-loki-bucket.yaml [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+
+      - name: Deploy Loki Zarf Package
+        working-directory: aws/loki
+        run: |
+          zarf package deploy zarf-package-*.tar.zst \
```
Comment (lucasrod16) [inline .github/workflows/test-loki-bucket.yaml:72]: We will need to add the `--confirm` flag here

**Review by lucasrod16 — CHANGES_REQUESTED** [review]

File: .github/workflows/test-loki-bucket.yaml (hunk 5 above)
Comment (UncleGedd) [reply .github/workflows/test-loki-bucket.yaml:72]: done!

File: aws/loki/variables.tf (hunk 2 above)
Comment (UncleGedd) [reply aws/loki/variables.tf]: done!

File: aws/loki/main.tf (hunk 1 above)
Comment (UncleGedd) [reply aws/loki/main.tf:29]: Fixed, using locals in that S3 module block now

File: .github/workflows/test-loki-bucket.yaml (hunk 3 above)
Comment (UncleGedd) [reply .github/workflows/test-loki-bucket.yaml:48]: love it, adding

File: aws/loki/zarf.yaml [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+    description: "Name of the DynamoDB table used for Terraform state locking"
+  - name: REGION
+    description: "The AWS region to run the Terraform in"
+
+  - name: LOKI_FORCE_DESTROY
+    description: "Provide a means to force destroy the Loki storage (S3) bucket even if it contains data"
+    default: "false"
+
+components:
+  - name: download-terraform
```
Comment (justin-o12) [inline aws/loki/zarf.yaml:32]: We had discussed moving this back to `aws/zarf.yaml` so other, future IaC components do not have to include this.  If we don't do it in this PR, we can refactor this as we add the other components (e.g. EKS).

**Review by justin-o12 — APPROVED** [review]
All LGTM.  We can refactor where we download Terraform if we decide it's necessary.

**Review by lucasrod16 — COMMENTED** [review]
@UncleGedd should we be implementing tests and zarf packaging for IaC modules in this repo?

It seems like we might benefit from doing those things in the source IaC repo, and only consume the published zarf package by importing the OCI package into our dubbd-aws package.

Thoughts?

**Comment by UncleGedd** [comment]
> @UncleGedd should we be implementing tests and zarf packaging for IaC modules in this repo?
> 
> It seems like we might benefit from doing those things in the source IaC repo, and only consume the published zarf package by importing the OCI package into our dubbd-aws package.
> 
> Thoughts?

Yeah I see where you're coming from. A few thoughts:
1. OCI skeletons [don't work as expected](https://github.com/defenseunicorns/zarf/issues/1808) for components involving files
2. The terratest tests are IMO akin to a "unit test." In this case it's testing that our DUBBD-AWS-specific Terraform configs are working as expected (ie. the the Terraform code in `aws/loki`)
3. I don't think we should necessarily put a `zarf.yaml` in the TF modules, at least not othis one. IMO the `zarf.yaml` that we import from `aws/loki` should contain opinionated DUBBD configuration, as opposed to the less opinionated configus exposed in the TF modules

Def open to more thoughts/feedback on this though!

File: aws/loki/zarf.yaml (hunk 6 above)
Comment (UncleGedd) [reply aws/loki/zarf.yaml]: moved!

**Review by justin-o12 — DISMISSED** [review]
LGTM

**Review by mjnagel — CHANGES_REQUESTED** [review]
CI is failing on `terraform output -raw irsa_role` in the AWS package (`Output "irsa_role" not found`) - I was trying to track down where/why and I think its related to moving `create_irsa` to a local?

**Comment by lucasrod16** [comment]
> > @UncleGedd should we be implementing tests and zarf packaging for IaC modules in this repo?
> > It seems like we might benefit from doing those things in the source IaC repo, and only consume the published zarf package by importing the OCI package into our dubbd-aws package.
> > Thoughts?
> 
> Yeah I see where you're coming from. A few thoughts:
> 
>     1. OCI skeletons [don't work as expected](https://github.com/defenseunicorns/zarf/issues/1808) for components involving files
> 
>     2. The terratest tests  in `aws/loki` are IMO akin to a "unit test." In this case it's testing that our DUBBD-AWS-specific Terraform configs are working as expected (ie. the the Terraform code in `aws/loki`)
> 
>     3. I don't think we should necessarily put a `zarf.yaml` in the TF modules, at least not this one. IMO the `zarf.yaml` that we import from `aws/loki` should contain opinionated DUBBD configuration, as opposed to the less opinionated configus exposed in the TF modules
> 
> 
> Def open to more thoughts/feedback on this though!

Thanks! Sounds good to me

**Comment by zachariahmiller** [comment]
@UncleGedd isn't it outputting null for the irsa role because you are passing in `--set create_irsa=false` in the zarf command? That would make sense to me. Regardless, this is maybe a little complicated for what you are actually aiming to accomplish.  I'm in support of making a generic irsa role anyways. Also, how would you feel about adding the IaC tests you are trying to do here as an example/loki tests/loki inside the s3 bucket module or to be created irsa module?

**Comment by UncleGedd** [comment]
> Regardless, this is maybe a little complicated for what you are actually aiming to accomplish.

Exactly this, the amount of workarounds I'm having to do to make this work feel off. 

 > adding the IaC tests you are trying to do here as an example/loki tests/loki
 
 I did add another test to the S3 module [here](https://github.com/defenseunicorns/terraform-aws-uds-s3/blob/98e72f23e1914b01012f751a213f7eda3c693cff/test/e2e/examples_test.go#L94), but if we refactor the IRSA into another module then I think this entire [commit](https://github.com/defenseunicorns/terraform-aws-uds-s3/commit/39a10229a7b2b4f2529d1ae1a72cec9b8a999f6e) could be reverted

**Comment by UncleGedd** [comment]
> outputting null for the irsa role because you are passing in --set create_irsa=false in the zarf command?

Yes, I want it to:
- default to true
- but be false when testing and running the `aws/loki` CI
- but be true when running the e2e tests with the cluster

Again, WAY too complicated, all to get around some AWS components that are tightly coupled

**Comment by zachariahmiller** [comment]
> > outputting null for the irsa role because you are passing in --set create_irsa=false in the zarf command?
> 
> Yes, I want it to:
> - default to true
> - but be false when testing and running the `aws/loki` CI
> - but be true when running the e2e tests with the cluster
> 
> Again, WAY too complicated, all to get around some AWS components that are tightly coupled and could be separated

Ah okay I think some of that context was lost on the mobile GitHub view. So yeah that would seem to require multiple zarf.configs to handle appropriately without adding multiple inputs and more testing logic. 

Only comment on the separate irsa module is that delivery is still consuming the s3 module that creates the irsa role so the implementation would need to still exist in the s3 bucket (call irsa module behind true/false) or would need to be a breaking change. I think this can be accomplished without issue and then just change the loki module to add irsa at the same level as the s3 bucket instead of nested inside, which _should_ facilitate what you are trying to accomplish. Still, if the tf output for the role name/arn is needed for the full deployment how do you plan on remedying that without getting into the same conditional rabbit hole?

**Comment by zachariahmiller** [comment]
But doesn't it still need to be conditionally called in the loki module (based on full cluster or not full cluster test)? And as a product of that be conditionally producing an output (which granted you could control to be like "" if it wasn't present) 

I'm all for what you're suggesting if you think you have a way around the current issue. Maybe tomorrow you can show me what you mean real quick and it will just click. In the meantime, I'm good with this immediate change in direction.

**Comment by UncleGedd** [comment]
man this is old!

**Pull request closed without merging**